### PR TITLE
Fix: update frame slider after deletion of layers

### PR DESCRIPTION
**Pull request by sfmig**
## Description

**What is this PR**

- [x] Bug fix
- [ ] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
If you load multiple files as points layers and delete one of them, the frame slider is automatically updated to reflect the dimensions of the remaining data. 

However, we currently only force the slider to match the full range of frames when a new point layer is loaded, not when one is removed. So after a delete event, if any file in the remaining data has all NaN values in the first frame (index `0`) or last frame (index `n`), napari will set the frame slider to start with `1` or end at `n-1` respectively. 

I use "frame slider" to refer to what napari documentation calls "dimension slider" for clarity, since in our widget it will represent number of frames most of the time.

**What does this PR do?**
This PR links the frame slider check to a "layer deletion" event. Specifically:
- It links layer deletion events to a `_on_layer_deleted` method, that runs `_check_frame_slider_range`
	- The check is skipped if there are no points layers in the viewer.
-  It adds a test that checks the frame slider is as expected when multiple files are loaded, one is deleted, and the remaining ones start or end with NaNs
- It adds a test to check there are no errors when all layers are deleted  

## References

Relates to #503  

## How has this PR been tested?

Tests pass locally and in CI.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

No.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [ n/a ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: tests/test_unit/test_napari_plugin/test_data_loader_widget.py [hunk 1]
```diff
@@ -302,10 +340,120 @@ def test_dimension_slider_multiple_files(
...
+                    .values,
+                ]
+            )
+            for ds in list_datasets
+        ]
+    )
+    assert actual_datasets_with_nans == expected_datasets_with_nans
+
+    # Get the maximum number of frames from all datasets
+    max_frames = max(ds.sizes["time"] for ds in list_datasets)
```
Comment (niksirbi) [inline tests/test_unit/test_napari_plugin/test_data_loader_widget.py:418]: this particular variable is never used, as this same line repeates itself a few lines down.

**Review by niksirbi — APPROVED** [review]
This also works as intended and I have nothing to comment on the implementation.

There are some strange (admittedly unlikely) corner cases though:
- let's say I load a short and a long dataset, each with its own video + Points layer
- I delete the longer Points layer, but I retain its video (Image layer). The changes introduced in this PR cause the slider to be updated according to the Points layer of the shorter dataset, meaning that parts of the longer video (whose Image layer is still loaded) are now inaccessible via the slider.

This will be more of a problem when we get to having both Points and Tracks layers for each movement dataset, but perhaps you already have a plan for that?

To say it in a different way, this PR establishes the primacy of the Points layer, i.e. the Points layers control the slider, ignoring other layer types. This may make sense for `movement`, but we have to decide what to do about corner cases like the one I describe above.

File: tests/test_unit/test_napari_plugin/test_data_loader_widget.py (hunk 1 above)
Comment (sfmig) [reply tests/test_unit/test_napari_plugin/test_data_loader_widget.py:425]: good catch, I think I actually missed an additional assertion I wanted to have in between - I added it.

**Comment by sfmig** [comment]
Thanks @niksirbi for having a look!

You raise a very good point, I made a few changes to accommodate for that:
- I simplified the callbacks connection (similar to the last addition to PR #503) 
- Now when we check if the range of the frame slider represents the full range of the loaded data, I inspect all loaded layers that are of type `Points`, `Tracks` or `Image`. If the layer does not have a `max_frame_idx` key in its metadata, I derive its maximum frame index from the shape of the data in the layer. That way the `Points` layer is not more important than any other layer that may have a concept of frame (the list of layers to consider can be updated in the future, e.g. to include `Shapes` layer if it defines shapes per frame).
- I added tests that:
	- verify that layers of different types are "attended to" by the frame slider update function, and
	- check the frame slider update function is called when a layer of any type is added or deleted. 
 Both make use of a fixture that defines sample data for each napari layer type.
- I have reorganised a bit the tests for clarity
